### ISSUE TITLE
CICIoT2023 Data Loading and Preprocessing

### ISSUE THREAD

**Issue by anupamme** [anupamme/iotsf_demo#6]
### Title
`[DATA] Implement CICIoT2023 dataset loader and preprocessor`

### Description
Create a data pipeline to load, preprocess, and serve samples from the CICIoT2023 dataset. The full dataset is 548GB, so we need a smart sampling strategy to extract representative samples for the demo. The loader should handle both benign traffic and various attack types.

### Acceptance Criteria
- [ ] Script to download a subset of CICIoT2023 (or instructions for manual download)
- [ ] Loader can read CSV/PCAP files from the dataset
- [ ] Preprocessor converts raw features to time-series format
- [ ] Can sample N benign and M attack samples of specific types
- [ ] Data is normalized and ready for model input
- [ ] Caching mechanism for preprocessed data
- [ ] Unit tests for data loading

### Technical Details

**CICIoT2023 Dataset Info:**
- Source: https://www.unb.ca/cic/datasets/iotdataset-2023.html
- 105 real IoT devices
- 33 attack types in 7 categories: DDoS, DoS, Recon, Web, BruteForce, Spoofing, Mirai
- Features include: flow duration, packet counts, byte counts, flag counts, etc.

**Key Features to Extract (12 dimensions for time-series):**
1. `flow_duration` - Duration of the flow
2. `fwd_pkts_tot` - Total forward packets
3. `bwd_pkts_tot` - Total backward packets
4. `fwd_data_pkts_tot` - Forward data packets
5. `bwd_data_pkts_tot` - Backward data packets
6. `fwd_pkts_per_sec` - Forward packets per second
7. `bwd_pkts_per_sec` - Backward packets per second
8. `flow_pkts_per_sec` - Flow packets per second
9. `fwd_byts_b_avg` - Average forward bytes per bulk
10. `bwd_byts_b_avg` - Average backward bytes per bulk
11. `fwd_iat_mean` - Forward inter-arrival time mean
12. `bwd_iat_mean` - Backward inter-arrival time mean

**src/data/loader.py:**
```python
"""
CICIoT2023 Dataset Loader

This module provides functionality to load and sample from the CICIoT2023 dataset.
The dataset is large (548GB), so we implement smart sampling strategies.
"""

import pandas as pd
import numpy as np
from pathlib import Path
from typing import Dict, List, Tuple, Optional
from loguru import logger

class CICIoT2023Loader:
    """
    Loader for CICIoT2023 dataset.
    
    The dataset contains network traffic from 105 IoT devices with 33 attack types.
    This loader handles:
    - Loading CSV files from the dataset
    - Filtering by attack type or benign traffic
    - Sampling representative subsets
    - Converting to time-series format
    """
    
    ATTACK_CATEGORIES = {
        'DDoS': ['DDoS-ICMP_Flood', 'DDoS-UDP_Flood', 'DDoS-TCP_Flood', 
                 'DDoS-PSHACK_Flood', 'DDoS-SYN_Flood', 'DDoS-RSTFINFlood',
                 'DDoS-SynonymousIP_Flood', 'DDoS-ICMP_Fragmentation',
                 'DDoS-ACK_Fragmentation', 'DDoS-UDP_Fragmentation',
                 'DDoS-HTTP_Flood', 'DDoS-SlowLoris'],
        'DoS': ['DoS-UDP_Flood', 'DoS-TCP_Flood', 'DoS-SYN_Flood', 
                'DoS-HTTP_Flood'],
        'Recon': ['Recon-PingSweep', 'Recon-OSScan', 'Recon-PortScan',
                  'Recon-HostDiscovery'],
        'Web': ['SqlInjection', 'CommandInjection', 'XSS',
                'BrowserHijacking', 'Backdoor_Malware', 'Uploading_Attack'],
        'BruteForce': ['DictionaryBruteForce'],
        'Spoofing': ['DNS_Spoofing', 'MITM-ArpSpoofing'],
        'Mirai': ['Mirai-greeth_flood', 'Mirai-greip_flood', 'Mirai-udpplain']
    }
    
    FEATURE_COLUMNS = [
        'flow_duration', 'fwd_pkts_tot', 'bwd_pkts_tot', 
        'fwd_data_pkts_tot', 'bwd_data_pkts_tot',
        'fwd_pkts_per_sec', 'bwd_pkts_per_sec', 'flow_pkts_per_sec',
        'fwd_byts_b_avg', 'bwd_byts_b_avg',
        'fwd_iat_mean', 'bwd_iat_mean'
    ]
    
    def __init__(self, data_dir: str, cache_dir: Optional[str] = None):
        """
        Initialize the loader.
        
        Args:
            data_dir: Path to raw CICIoT2023 CSV files
            cache_dir: Path to cache preprocessed data (optional)
        """
        self.data_dir = Path(data_dir)
        self.cache_dir = Path(cache_dir) if cache_dir else None
        self._validate_data_dir()
        
    def _validate_data_dir(self):
        """Check if data directory exists and contains expected files."""
        if not self.data_dir.exists():
            logger.warning(f"Data directory {self.data_dir} does not exist. "
                          f"Run scripts/download_data.py first.")
    
    def load_benign_samples(self, n_samples: int, 
                            seq_length: int = 128) -> np.ndarray:
        """
        Load benign traffic samples.
        
        Args:
            n_samples: Number of sequences to return
            seq_length: Length of each time-series sequence
            
        Returns:
            Array of shape (n_samples, seq_length, n_features)
        """
        # Implementation: Load benign traffic, segment into sequences
        pass
    
    def load_attack_samples(self, attack_type: str, n_samples: int,
                           seq_length: int = 128) -> np.ndarray:
        """
        Load attack traffic samples of a specific type.
        
        Args:
            attack_type: One of the attack types from ATTACK_CATEGORIES
            n_samples: Number of sequences to return
            seq_length: Length of each time-series sequence
            
        Returns:
            Array of shape (n_samples, seq_length, n_features)
        """
        # Implementation: Load attack traffic, segment into sequences
        pass
    
    def get_mixed_batch(self, n_benign: int, n_attack: int,
                        attack_types: List[str],
                        seq_length: int = 128) -> Tuple[np.ndarray, np.ndarray]:
        """
        Get a mixed batch of benign and attack samples.
        
        Returns:
            Tuple of (samples, labels) where labels are 0 for benign, 1 for attack
        """
        pass
```

**src/data/preprocessor.py:**
```python
"""
Data Preprocessor for IoT Network Traffic

Converts raw network flow features into normalized time-series format
suitable for diffusion models and anomaly detection.
"""

import numpy as np
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from typing import Tuple, Optional
import pickle
from pathlib import Path

class TrafficPreprocessor:
    """
    Preprocessor for network traffic data.
    
    Handles:
    - Feature normalization (StandardScaler or MinMaxScaler)
    - Sequence creation from flow records
    - Train/test splitting
    - Scaler persistence for inference
    """
    
    def __init__(self, scaler_type: str = 'standard'):
        """
        Initialize preprocessor.
        
        Args:
            scaler_type: 'standard' for StandardScaler, 'minmax' for MinMaxScaler
        """
        self.scaler_type = scaler_type
        self.scaler = None
        self._fitted = False
        
    def fit(self, data: np.ndarray) -> 'TrafficPreprocessor':
        """
        Fit the scaler on training data.
        
        Args:
            data: Array of shape (n_samples, n_features) or 
                  (n_samples, seq_length, n_features)
        """
        # Flatten if 3D
        if data.ndim == 3:
            data = data.reshape(-1, data.shape[-1])
            
        if self.scaler_type == 'standard':
            self.scaler = StandardScaler()
        else:
            self.scaler = MinMaxScaler()
            
        self.scaler.fit(data)
        self._fitted = True
        return self
    
    def transform(self, data: np.ndarray) -> np.ndarray:
        """
        Transform data using fitted scaler.
        
        Args:
            data: Array of shape (n_samples, n_features) or
                  (n_samples, seq_length, n_features)
                  
        Returns:
            Normalized data of same shape
        """
        if not self._fitted:
            raise RuntimeError("Preprocessor not fitted. Call fit() first.")
            
        original_shape = data.shape
        if data.ndim == 3:
            data = data.reshape(-1, data.shape[-1])
            
        normalized = self.scaler.transform(data)
        
        if len(original_shape) == 3:
            normalized = normalized.reshape(original_shape)
            
        return normalized
    
    def inverse_transform(self, data: np.ndarray) -> np.ndarray:
        """Inverse transform normalized data back to original scale."""
        pass
    
    def save(self, path: str):
        """Save fitted scaler to disk."""
        with open(path, 'wb') as f:
            pickle.dump(self.scaler, f)
            
    def load(self, path: str):
        """Load fitted scaler from disk."""
        with open(path, 'rb') as f:
            self.scaler = pickle.load(f)
        self._fitted = True


def create_sequences(data: np.ndarray, seq_length: int, 
                     stride: int = 1) -> np.ndarray:
    """
    Convert raw flow records into overlapping sequences.
    
    Args:
        data: Array of shape (n_records, n_features)
        seq_length: Desired sequence length
        stride: Step size between sequences
        
    Returns:
        Array of shape (n_sequences, seq_length, n_features)
    """
    n_records, n_features = data.shape
    n_sequences = (n_records - seq_length) // stride + 1
    
    sequences = np.zeros((n_sequences, seq_length, n_features))
    for i in range(n_sequences):
        start_idx = i * stride
        sequences[i] = data[start_idx:start_idx + seq_length]
        
    return sequences
```

**scripts/download_data.py:**
```python
"""
Download a subset of CICIoT2023 for demo purposes.

The full dataset is 548GB. This script downloads only the necessary
subset for the demo (approximately 500MB).
"""

import os
import requests
from pathlib import Path
from tqdm import tqdm
import zipfile

# CICIoT2023 is hosted on UNB's servers
# Note: You may need to manually download due to access restrictions

DATASET_INFO = """
CICIoT2023 Dataset Download Instructions:
==========================================

1. Visit: https://www.unb.ca/cic/datasets/iotdataset-2023.html
2. Fill out the registration form
3. Download the following files for the demo:
   - Part 1 (contains benign traffic)
   - Part 3 (contains Mirai and DDoS)
   - Part 5 (contains reconnaissance)

4. Extract to: data/raw/ciciot2023/

Alternative: Use the provided sample data in data/sample/ for testing.
"""

def download_sample_data(output_dir: str):
    """
    Download sample data for testing (synthetic, not real CICIoT2023).
    This is for development/testing when full dataset isn't available.
    """
    output_path = Path(output_dir)
    output_path.mkdir(parents=True, exist_ok=True)
    
    # Generate synthetic sample data for development
    import numpy as np
    import pandas as pd
    
    n_samples = 10000
    n_features = 12
    
    feature_names = [
        'flow_duration', 'fwd_pkts_tot', 'bwd_pkts_tot',
        'fwd_data_pkts_tot', 'bwd_data_pkts_tot',
        'fwd_pkts_per_sec', 'bwd_pkts_per_sec', 'flow_pkts_per_sec',
        'fwd_byts_b_avg', 'bwd_byts_b_avg',
        'fwd_iat_mean', 'bwd_iat_mean'
    ]
    
    # Generate benign traffic (normal distribution)
    benign_data = np.random.randn(n_samples, n_features) * 0.5 + 1
    benign_df = pd.DataFrame(benign_data, columns=feature_names)
    benign_df['label'] = 'Benign'
    
    # Generate attack traffic (shifted distribution)
    attack_data = np.random.randn(n_samples, n_features) * 0.8 + 1.5
    attack_df = pd.DataFrame(attack_data, columns=feature_names)
    attack_df['label'] = 'DDoS-TCP_Flood'
    
    # Save
    benign_df.to_csv(output_path / 'benign_sample.csv', index=False)
    attack_df.to_csv(output_path / 'attack_sample.csv', index=False)
    
    print(f"Sample data saved to {output_path}")


if __name__ == '__main__':
    print(DATASET_INFO)
    
    response = input("Generate synthetic sample data for development? [y/N]: ")
    if response.lower() == 'y':
        download_sample_data('data/raw/sample')
```

### Dependencies
- Issue #1 (Project Setup)

### Estimated Effort
4-5 hours

### Labels
`data`, `preprocessing`, `priority-high`
